torch/distributed/fsdp/_shard_utils.py

```diff
@@ -124,3 +141,28 @@ def _create_chunk_sharded_tensor(
     return ShardedTensor._init_from_local_shards_and_global_metadata(
         local_shards, sharded_tensor_metadata=sharded_tensor_metadata, process_group=pg
     )
+
+
+def _create_chunk_dtensor(
+    tensor: torch.Tensor,
+    rank: int,
+    device_mesh: DeviceMesh,
+) -> DTensor:
+    """
+    Shard a tensor to chunks along the first dimension. The local rank will gets its
+    corresponding chunk as the local tensor to create a DTensor.
+    """
+    shard_placement = DShard(0)
+    tensor_padded_list, pad_sizes = shard_placement._split_tensor(
+        tensor,
+        device_mesh.size(dim=0),
+        with_padding=True,
+        contiguous=True,
+    )
+    unpadded_list = [
+        shard_placement._unpad_tensor(tensor, pad_sizes[i])
+        if pad_sizes[i] > 0
+        else tensor
+        for i, tensor in enumerate(tensor_padded_list)
+    ]
+    return DTensor.from_local(unpadded_list[rank], device_mesh, [shard_placement])
```

Comment:
I'm not familiar with DTensor. Does this involve any communication? If so, it will be expensive for state_dict.

Reply:
From my understanding, from_local does not involve any communication as long as it's just shard. Only replicate on a dimension would require communication.  

Code pointer: 
https://github.com/pytorch/pytorch/blob/main/torch/distributed/_tensor/api.py#L243